Use file content heuristics to decide file reader.

Dimi1010 opened the pull request:
The PR adds heuristics based on the file content that is more robust than deciding based on the file extension.

The new decision model scans the start of the file for its magic number signature. It then compares it to the signatures of supported file types [[1]](https://en.wikipedia.org/wiki/List_of_file_signatures) and constructs a reader instance based on the result.

A new function `createReader` and `tryCreateReader` has been added due to changes in the public API of the factory.
The functions differ in the error handling scheme, as `createReader` throws and `tryCreateReader` returns `nullptr` on error.

Method behaviour changes during erroneous scenarios:
| Scenario | `getReader` | `createReader` | `tryCreateReader` |
| ------------- | ------------- | ------------- | ------------- |
| File not found | N/A | Throws exception | Return `nullptr` |
| Unsupported format | Return `PcapFileDeviceReader` | Throws exception | Return `nullptr` |

Dimi1010 committed:
Added heuristics file content detector that determines the content based on the magic number.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Moved stream checkpoint outside format detector as it is not directly tied to it.

Dimi1010 committed:
Added a new factory function `createReader` that uses the new heuristics detection method.

Dimi1010 committed:
Add <algorithm> include.

Dimi1010 committed:
Added unit tests.

Dimi1010 committed:
Deprecated old factory function.

Dimi1010 committed:
Add byte-swapped zstd magic number.

Dimi1010 committed:
Lint

Dimi1010 committed:
Move enum closer to first usage.

Dimi1010 commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -28,32 +30,186 @@ namespace pcpp
+				return std::find(pcapMagicNumbers.begin(), pcapMagicNumbers.end(), magic) != pcapMagicNumbers.end();
+			}
+
+			bool isSnoopFile(std::istream& content)
```
ditto: https://github.com/seladb/PcapPlusPlus/pull/1962#discussion_r2348771886

Dimi1010 committed:
Added unit tests for file reader device factory.

Dimi1010 commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -763,26 +763,57 @@ PTF_TEST_CASE(TestPcapNgFileReadWriteAdv)
-	PTF_ASSERT_NOT_NULL(dynamic_cast<pcpp::PcapNgFileReaderDevice*>(genericReader));
-	PTF_ASSERT_TRUE(genericReader->open());
+	// ------- IFileReaderDevice::createReader() Factory
+	// TODO: Move to a separate unit test.
```
3d713ab3457ad3c7012e1f79b3480b469ea664d8 adds the following tests:
- Pcap, PcapNG, Zst file with correct content + extension
- Pcap, PcanNG file with correct content + wrong extension
- Bogus content file with correct extension (pcap, pcapng, zst)
- Bogus content file with wrong extension (txt)

Haven't found a snoop file to add. Do we have any?

> Open pcap files with different magic numbers

Do you mean Pcap content that has just its magic number changed? Because IMO it is reasonable to consider that invalid format and fail as regular bogus data.

> Assuming we add a version check for snoop and pcap file: create temp files with bogus data that has the magic number but wrong versions

Pending on https://github.com/seladb/PcapPlusPlus/pull/1962#discussion_r2348771886 .

Dimi1010 committed:
Revert indentation.

Dimi1010 committed:
Fixed StreamCheckpoint to restore original stream state.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -28,32 +30,189 @@ namespace pcpp
+		private:
+			bool isPcapFile(std::istream& content)
+			{
+				constexpr std::array<uint32_t, 4> pcapMagicNumbers = {
+					0xa1'b2'c3'd4,  // regular pcap, microsecond-precision
+					0xd4'c3'b2'a1,  // regular pcap, microsecond-precision (byte-swapped)
+					0xa1'b2'3c'4d,  // regular pcap, nanosecond-precision
+					0x4d'3c'b2'a1   // regular pcap, nanosecond-precision (byte-swapped)
+				};
```
Add KUZNETZOV_TCPDUMP_MAGIC as it is supported by libpcap. [[1]](https://github.com/the-tcpdump-group/libpcap/blob/56fbeb87e60818a98368f0c2f04bec796935e0bd/sf-pcap.c#L238-L248)

Dimi1010 replied:
Added 916e8728f6b3a2a20c16e64d874f1396537744c9

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -28,32 +30,186 @@ namespace pcpp
+			}
+
+		private:
+			bool isPcapFile(std::istream& content)
```
ok sure, we can leave just the magic number

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Moved isStreamSeekable helper to inside `CaptureFileFormatDetector`.

Move it out if it needs to be reused somewhere.

Dimi1010 committed:
Added pcap magic number for Alexey Kuznetzov's modified pcap format.

Libpcap supports reading this format since 0.9.1. The heuristics detection will identify such magic number as pcap and leave final support decision to the pcap backend infrastructure.

seladb commented:
@Dimi1010 some CI tests fail...

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Split the unit test into multiple smaller tests.

seladb commented:
> > @Dimi1010 some CI tests fail...
> 
> @seladb I think I found the issue. According to ChatGPT the Winpcap's `wpcap.dll` does not actually support nanosecond precision files [1], due to being based on libpcap 1.0.x. Nanosecond precision pcaps were added in libpcap 1.5.
> 
> The NPcap sdk and drivers provide their own wpcap.dll where they do support nanosec precision.
> 
> The tests we have on nanosecond precision pcap files ("nanosecs.pcap") currently are flawed. We don't use a static base base to compare to. We write a packet with the `PcapFileWriterDevice` configured to write nanoseconds precision and then attempt to read that with `PcapFileReaderDevice`. The issue with that is that the writer has 0 error reporting if its configured to write nanosecond precision but it can't due to `PCAP_TSTAMP_PRECISION_NANO` not being defined and just uses `pcap_open_dead` which uses default precision.
> 
> Due to the above the current nanosecond test in the master branch passes with Winpcap, even though it doesn't actually write / read a nanosecond precision pcap.
> 
> The new tests use a static copy of `nanoseconds.pcap` generated from Npcap for base image to compare to. Due to that, the a `PcapFileWriterDevice` with the `Winpcap` backend fails to read that and fails the test due to `open()` failure.
> 
> Edit: [1] - Confirmed that by looking at the Winpcap DLL API. It does not have functions like `pcap_open_dead_with_tstamp_precision`, etc.

Maybe we can remove pcap files that have nanoseconds precision from this PR and replace them with other pcap files to make the tests pass?

Dimi1010 commented:
> Maybe we can remove pcap files that have nanoseconds precision from this PR and replace them with other pcap files to make the tests pass?

Then, we don't have coverage for the magic number for nanosecond pcap.

See #1977. It is a fix for the underlying issue.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

# Conflicts:
#	Pcap++/src/PcapFileDevice.cpp

Dimi1010 committed:
Added helper to indicate if ZstSupport is enabled for PcapNg devices.

Dimi1010 committed:
Split pcap microsecond and nanosecond file heuristics tests.
Nanoseconds are skipped if they aren't supported.

Dimi1010 committed:
Skipping Zst test case if zst is not supported.

Dimi1010 committed:
Due to file heuristics returning PcapNG format on Zstd archive, if Zstd compression is unsupported, the format detector will return Unknown.

Dimi1010 committed:
Lint

Dimi1010 committed:
Added invalid device factory to pcap tag.

Dimi1010 committed:
Updated static zst archives to be actual archives.

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -29,43 +53,195 @@ namespace pcpp
+
+				// PcapNG backend can support ZstdCompressed Pcap files, so we assume an archive is compressed PcapNG.
+				// If Zstd is not supported, we cannot read the file anyway, so we return Unknown.
+				if (isPcapNgFile(content) || (checkZstdSupport() && isZstdArchive(content)))
```
Yeah, I agree that it's the correct behavior 👍

seladb commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,125 @@ class FileReaderTeardown
+
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Invalid)
+{
+	// Garbage data, correct extension
+	constexpr const char* PCAP_BOGUS_FILE_PATH = "PcapExamples/file_heuristics/bogus-content.pcap";
+	constexpr const char* PCAPNG_BOGUS_FILE_PATH = "PcapExamples/file_heuristics/bogus-content.pcapng";
+	constexpr const char* PCAPNG_ZST_BOGUS_FILE_PATH = "PcapExamples/file_heuristics/bogus-content.zst";
+
+	// Garbage data
+	constexpr const char* BOGUS_FILE_PATH = "PcapExamples/file_heuristics/bogus-content.txt";
```
Ditto: since the extension doesn't matter, do we need all 4 files?

seladb replied:
I don't think it's really needed...

Dimi1010 replied:
Updated 07804dac2cb388e903823693b6c16baf84be4ca9
Left only the .txt case, since the others are duplicates.

Dimi1010 commented on Tests/Pcap++Test/TestDefinition.h:
```diff
@@ -20,6 +20,12 @@ PTF_TEST_CASE(TestLogger);
 PTF_TEST_CASE(TestLoggerMultiThread);
 
 // Implemented in FileTests.cpp
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Pcap_MicroPrecision);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Pcap_NanoPrecision);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG_ZST);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG_ZST_Unsupported);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Invalid);
```
I don't like merging them all into one test since they have different skip conditions, and having them together will make that logic convoluted.
- NanoPrecision gets skipped on non-nano environments.
- Zst gets skipped on non-zst environments.
- Zst Unsupported gets skipped on zst environments.

IMO, smaller test cases are more readable and make it easier to identify the error than one large test block that tries to have everything together. I don't think the compilation time increase will be noticeable.

seladb replied:
Ok, I guess we can leave them separate

Dimi1010 commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,125 @@ class FileReaderTeardown
+		PTF_SKIP_TEST("Zstandard compression is not supported in this platform/environment");
+	}
+
+	constexpr const char* PCAPNG_ZST_FILE_PATH = "PcapExamples/file_heuristics/pcapng-example.pcapng.zst";
+	constexpr const char* PCAPNG_ZSTD_FILE_PATH = "PcapExamples/file_heuristics/pcapng-example.pcapng.zstd";
```
Fair enough. I will remove the zstd file. I think the .zst is the more common extension?

Dimi1010 commented on Tests/Pcap++Test/TestDefinition.h:
```diff
@@ -20,6 +20,12 @@ PTF_TEST_CASE(TestLogger);
 PTF_TEST_CASE(TestLoggerMultiThread);
 
 // Implemented in FileTests.cpp
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Pcap_MicroPrecision);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Pcap_NanoPrecision);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG_ZST);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_PcapNG_ZST_Unsupported);
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Invalid);
```
Do we have a reason to stick to 35 chars output? Perhaps we can up it to something like 60~80?
I can reduce the names, but it would hurt the specificity of the test name.

Dimi1010 replied:
Somewhat shortened them, but also increased the width to 45.
Should be fine now?

4f52f59a948b6a24beacc26ef0413e351c40c4ef & 41fe18845a271c40f1f1347e23cba87d2f96e793

Dimi1010 committed:
Centralized PTF test name width under a macro.

Dimi1010 committed:
Add Pcap++Test header files to test sources for IDE tooling.

Dimi1010 committed:
Fixed test output formatting.

Dimi1010 committed:
Lint

Dimi1010 committed:
Typo fix.

Dimi1010 commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,125 @@ class FileReaderTeardown
+{
+	// Correct format
+	constexpr const char* PCAP_MICROSEC_FILE_PATH = "PcapExamples/file_heuristics/microsecs.pcap";
+	// Correct format, wrong exteension, microsecond precision
```
Fixed c7cab2b4835006c48b3f0d22ae442d789a0a08e0

seladb commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,125 @@ class FileReaderTeardown
 	}
 };
 
+PTF_TEST_CASE(TestIFileReaderDeviceFactory_Pcap_MicroPrecision)
```
I mean the other possible magic numbers of a valid pcap file. Since it's not easy to find such pcap files, we can generate synthetic files that are not actually valid, but will look valid for the sake of the test

seladb replied:
I'd suggest we have spoofed pcap samples for all options

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Shortened test names.

Dimi1010 committed:
Simplified invalid file test.

Dimi1010 committed:
Simplified ZST tests.

Dimi1010 committed:
Added snoop test.

Dimi1010 committed:
Expanded granularity of file format detection.

Updated pcap file detection to return the precice format of Pcap instead of just `true` / `false`.
Updated detect format to always retuirn the detected format. Previous responsibility for unsupported zstd archive files has been passed up the call stack to the factory function `createReader`.

Dimi1010 committed:
Marked `checkSupport` functions as constexpr to enable compile time optimizations and branch pruning.

Dimi1010 committed:
Exclude json from pre-commit cppcheck as it is slow due to many define branches.

Dimi1010 committed:
Lint

Dimi1010 committed:
Fix runtime side effects inside constexpr function.

Dimi1010 commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -131,6 +336,45 @@ namespace pcpp
 	}
 
+	std::unique_ptr<IFileReaderDevice> IFileReaderDevice::createReader(const std::string& fileName)
+	{
+		std::ifstream fileContent(fileName, std::ios_base::binary);
+		if (fileContent.fail())
+		{
+			throw std::runtime_error("Could not open: " + fileName);
+		}
+
+		switch (CaptureFileFormatDetector().detectFormat(fileContent))
+		{
+		case CaptureFileFormat::PcapNano:
+		{
+			if (!checkNanoSupport())
+			{
+				PCPP_LOG_ERROR("Pcap nano second precision files are not supported in this build of PcapPlusPlus");
+				return nullptr;
+			}
+			// fallthrough
+		}
+		case CaptureFileFormat::Pcap:
+			return std::make_unique<PcapFileReaderDevice>(fileName);
+		case CaptureFileFormat::PcapNGZstd:
+		{
+			if (!checkZstdSupport())
+			{
+				PCPP_LOG_ERROR("PcapNG Zstd compressed files are not supported in this build of PcapPlusPlus");
+				return nullptr;
+			}
+			// fallthrough
+		}
+		case CaptureFileFormat::PcapNG:
+			return std::make_unique<PcapNgFileReaderDevice>(fileName);
+		case CaptureFileFormat::Snoop:
+			return std::make_unique<SnoopFileReaderDevice>(fileName);
+		default:
+			return nullptr;
+		}
+	}
```
Curious if it isn't simpler to throw on unsupported format, instead of returning `nullptr`... It would give the caller only 1 error mechanism to worry about.

Generally those are unrecoverable errors for the context, and we already throw on missing file?
If needed we can provide a second `tryCreate` factory for non-exceptional error flow.

Dimi1010 replied:
Updated 45ad7695f0b73de6ccc3669666d48c6985523c1f
`tryCreateReader` has a dirty implementation, but it is not supposed to be used in performance heavy code. How often do people create new devices?

seladb replied:
I wouldn't throw an exception for unsupported format. Unlike other languages, using `try..catch` in C++ isn't very common, mostly because it's a costly operation. Returning `nullptr` is better in this case

Dimi1010 committed:
Added a secondary factory function to separate mixed error handling methods.

Dimi1010 committed:
Revert deprecation message, as doxygen is unhappy.

Dimi1010 committed:
Update tests.

Dimi1010 committed:
Update deprecation warning to point to the function closer to the signature.

Dimi1010 committed:
Catch general exception instead of runtime error.

Dimi1010 committed:
Shortened deprecation message due to pre-commit warnings when its is on 1 line and doxygen errors when its in 2 lines.

Dimi1010 committed:
Fix braces.

Dimi1010 committed:
Simplfy test.

Dimi1010 committed:
Added tests for createReader failures.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Simplified pcap detection to not require to read the entire pcap header.

Dimi1010 committed:
Added const qualifiers to detector methods.

Dimi1010 committed:
Added dedicated unit tests for CaptureFileFormatDetector.

Dimi1010 committed:
Added more tests for `createReader`.

Dimi1010 committed:
Add static assert for array indice checks.

seladb commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,204 @@ class FileReaderTeardown
 	}
 };
 
+PTF_TEST_CASE(TestFileFormatDetector)
```
Please see my previous comment. Maybe we can create a temp fake file with the expected data and run `createReader()`

Dimi1010 commented on Pcap++/header/PcapFileDevice.h:
```diff
@@ -19,6 +19,36 @@ namespace pcpp
+		};
+
+		/// @brief Heuristic file format detector that scans the magic number of the file format header.
+		class CaptureFileFormatDetector
```
I tried that suggestion initially, but it would have been an extremely fragile unit test. The "pass" conditions would have been checked indirectly.

Also, `createReader` has multiple return paths for Nano / Zst file formats, which would have caused complications since the format test would have needed to care about the environment it runs at, which it doesn't have to as a standalone.

Any additional changes to `createReader` could also break the test, which they really shouldn't. For example, I am thinking of maybe adding additional logic for Zst archive to check if the compressed data is actually a pcapng, and not a random file. This would be a nightmare to make compatible with the "spoofed files" test due to assumptions on the test that `createReader` doesn't do anything more complicated than check the initial magic number.

So, in the end, you end up with a more compilcated unit test to read through that: 
- depends on the environment it runs on.
- can be broken not just by changes to the format detector but also changes to the `createReader` factory, too.
- induces requirements on `createReader` as it uses its behavior to test `detectFormat`.

seladb replied:
> It is a big difference and it's not always an easy fix. I plan to add the aforementioned Zst checks in another PR after this one, and that would make zst spoofing in `createReader` impossible, due to zst format automatically being checked for PcapNg or Unknown contents. Therefor you can't rely on the return of `createReader` to find out what the return of `detectFormat` was, because `nullptr` can be returned from several paths from `detectFormat` return value (Unknown, Nano + unsupported, Zst + unsupported). We have already had issues with tests being silently broken (#1977 comes to mind), so I would prefer to avoid fragile tests if we can.

I'm not sure I understand... if we create fake files we know which type to expect, so all the test needs to do is verify the created file device is of the expected type 🤔 

> * In the current case, we have another option. Since the `CaptureFileFormatDetector` is only needed in the `cpp` part and not in the header part, we can extract it to a fully internal header, kept with the source files. This would prevent it from being exposed in the public API, but the Test project can be manually set to search for headers from "Pcap++/src" too, to allow it to link in the tests.

I guess we can do that, but I still don't understand why we can't test it with `createReader` or `tryCreateReader`

Dimi1010 replied:
d8f741966dc25db955e4ab9ccd7ec9a405116c25 Updated to separate header / source under Pcap++/src

seladb replied:
@Dimi1010 I understand all of that, but the scenarios we want to test are positive scenarios, which means that if `createReader` return `nullptr` on a "valid" fake file, it's probably a bug. What you describe might be more relevant if the test expects a `nullptr` and it's harder to pin-point which branch in exactly it `createReader` came from.

Please let me know if I'm missing something here

Dimi1010 replied:
> > Tbh, I think that unit tests should mostly validate public API for an object. Not `private`, as that is compilated.
> 
> > For me what guides me is the Single Responsibility Principle:
> > ```
> > * `createReader` factory has the sole responsibility of creating the correct reader for a given file format.
> > 
> > * `CaptureFileFormatDetector::detectFormat` has the sole responsibility of determining the file format of a given byte stream.
> > ```
> 
> These 2 statements sometimes contradict, and this is a good example: on one hand `createReader` is the public API which you claim we should test, while `CaptureFileFormatDetector` is considered private and shouldn't be tested, on the other hand, the Single Responsibility Principle dictates testing this "private" code. However, since C++ doesn't allow accessing private/protected methods, I usually examine it case-by-case. In this case I think it's ok to only test `createReader` because it covers `CaptureFileFormatDetector` well
> 

Sometimes they can, yes, but key word here is "**for an object**". What is public depends on the context you are looking at:
1. If you are looking at the library as a whole, what is considered public is what is in the public headers. The rest are internal headers.
2. If you are looking at a specific class, what is considered public is the class's public methods.

Generally, unit tests should validate code behaviour in the most isolated case that can be achieved (that isn't a trivial case) to keep them simple and so you can easily **locate the error**. That means that cases should be isolated enough that, in the case of failure to show the smallest reasonable region of where the error is, not just that the error exists (example below). As such, the context most looked at should be option 2 as that generally keeps the individual tests more isolated and on the smaller size. The format detector's logic is complex enough that it justifies for as a separate unit.

> [...] while `CaptureFileFormatDetector` is considered private and shouldn't be tested, on the other hand, the Single Responsibility Principle dictates testing this "private" code.

Under option 2, `CaptureFileFormatDetector` is not considered "private" but a separate object that must be validated independently as it is not trivial.

> However, since C++ doesn't allow accessing private/protected methods, I usually examine it case-by-case.

In this case, we have no such issues, as the format detector is a separate reusable object, and not a private method of `IFileReaderDevice`. The only issue is that of linkage, as the tests should be able to link to the format detector. C++ allows that naturally by having it as external linkage, but you don't want it exposed to the public API (as it is rightfully not needed for the users), therefor separate header that is considered "internal", not "private".

In the current case "I think that unit tests should mostly validate public API for an object" means that the format detector should only have a unit test for `detectFormat` and not for all the implementations of private methods `detectPcap`, `isPcapNG`, etc.

Here is an example:
```cpp

// This should be tested in a standalone case with respect to Bar unit tests.
// Assume this is not a trivial computation.
int bar(int a) { 
  if (a < 0)
    return 0;
  return a + 42;
}

class Foo
{
public:
  // This should be tested in a standalone case with respect to Foo unit tests.
  int doFoo(int a) {
    if (validateA(a))
    {
      return doFooImpl(a);
    }
    return 0;
  }
  
private:
  // This should NOT be tested in a standalone case, as it is considered "private" implementation.
  int doFooImpl(int a) {
    auto barVal = bar(a);
    if (barVal > 50)
    {
        return barVal + 5;
    }
    return barVal + 10;
  }
  
  // This should NOT be tested in a standalone case, as it is considered "private" implementation  
  bool validateA(int a) { 
    return a >= 5;
  }
}
```

In the above case, you have 2 logical independent reusable units: Bar and Foo.
Foo's unit tests should concert themselves only with validating the input / output behaviour of `doFoo` and other public methods of `Foo`. It should not concern itself with validating `bar()`'s behaviour, because:
- Testing `bar()` is testing the precise internal implementation of `doFoo()`, instead of just validating externally observable behaviour.
- It **can't** do it cleanly: `doFoo(...)`'s valid range is [5, IntMax], while `bar(...)`'s valid range is [0, IntMax]. How do you test the other values?

For the aforementioned statement:
> That means that cases should be isolated enough that, in the case of failure to show the smallest reasonable region of where the error is, not just that the error exists (example below).

If `Foo` tests fail, but `Bar` tests pass, that means that the error is in Foo, if only `Foo` tests existed, then we would need to also go look through `bar(...)` since we wouldn't know that it runs fine.

In summary:
- `Foo`'s unit tests should treat everything under `Foo::doFoo(int a)` as a black box, and only validate the externally observable behaviour.
- In its turn, `Foo` should treat `bar(...)` as a black box that does what it says on the tin can.
- Bar's unit tests should validate the full input / output behaviour of `bar(int a)` (so it actually does what it says it does) and treat `bar`'s implementation of how it achieves that as a black box.

Now replace `Foo::doFoo` with `createReader` and `bar` with `detectFormat`.

> In this case I think it's ok to only test `createReader` because it covers `CaptureFileFormatDetector` well

It does not cover it well, considering we would need to resort to workarounds such as creating nominally **invald** fake files that should realistically be discarded by the factory validation, as they can't produce a valid reader device.

The format detector might work with them, because of its current implementation only scans the first 4 to 6 bytes, but that does not mean that the reader factory should.

Realistically, after the format is determined and a device is created, the factory should attempt to open the file to read it to validate it can before returning the reader (and optionally closing it). It would not be able to do that, if it has to accept those fake files for the purposes of testing something that should be considered a black box that "just works" from its perspective.  Allowing those "fake" files locks us out of any other potential validation we might be able to do on the files, when there are cleaner solutions that don't lock us out of it. That or we now need to modify the fake files again, which is also more work that does not happen with separate unit tests.

> `ObjectPool` is different because it's not specific to the `Logger` and can be reused elsewhere. `CaptureFileFormatDetector` is very specific to this file, so extracting it to its own `.h` and `.cpp` files just for the sake of testing seems redundant to me

I agree that it was specifically made **for** the factory's needs, but not that it is specific **to** it.

The format detector is perfectly viable to be viewed as a black box from the factory's perspective. File contents go in, format enum value goes out. That same black box can be reused elsewhere if needed, making it independent from the factory's implementation.

seladb replied:
> Generally, unit tests should validate code behaviour in the most isolated case that can be achieved (that isn't a trivial case) to keep them simple and so you can easily **locate the error**. That means that cases should be isolated enough that, in the case of failure to show the smallest reasonable region of where the error is, not just that the error exists (example below). As such, the context most looked at should be option 2 as that generally keeps the individual tests more isolated and on the smaller size. The format detector's logic is complex enough that it justifies for as a separate unit.

I don't disagree on the concept, however the tradeoff I'm trying to make is between:
1. Refactor code just for the sake of writing tests
2. Be able to write tests
3. Write tests that are isolated enough, as you mentioned

If (2) or (3) are impossible, then we may consider (1), but if they are feasible, even if not perfect, we should try to avoid (1). For example: if I'd re-write the DPDK device, I'd probably use dependency injection so we can mock the DPDK API and test at least some of our business logic.

If the code-under-test is completely isolated from the business logic (for example: `ObjectPool`), then (1) is not true, and it makes sense to move this code to separate header/source files.

> Under option 2, `CaptureFileFormatDetector` is not considered "private" but a separate object that must be validated independently as it is not trivial.

Yes, you're right, I'm sorry for the inaccurate wording. What I meant is that `CaptureFileFormatDetector` is not part of the public API, but rather an internal class



> In this case, we have no such issues, as the format detector is a separate reusable object, and not a private method of `IFileReaderDevice`. The only issue is that of linkage, as the tests should be able to link to the format detector. C++ allows that naturally by having it as external linkage, but you don't want it exposed to the public API (as it is rightfully not needed for the users), therefor separate header that is considered "internal", not "private".

Again - the main thing that guides me is: do we need this refactoring just for the sake of testing? The answer in this case is "yes".

Regarding the example you gave: if `bar` is completely isolated from `Foo` and may be reusable in other places (like `ObjectPool`), then sure - we can move `bar` to a separate header/source file. But if `bar` is an internal implementation of `Foo`, and we have a good enough way to test `doFooImpl` / `bar` using the public method `doFoo` then I'd go with this approach

Dimi1010 replied:
> If (2) or (3) are impossible, then we may consider (1), but if they are feasible, even if not perfect, we should try to avoid (1). For example: if I'd re-write the DPDK device, I'd probably use dependency injection so we can mock the DPDK API and test at least some of our business logic.

Sure, I don't disagree that refactoring just for writing tests for existing features and I am not suggesting a rewrite of existing code or tests. 

I think our main disagreement is what is considered an existing feature? I consider the format detector a **new** feature, independent of the `createReader` factory's business logic. It might be internal to the library but it is still separate from other existing features.

I view the PR as the following sequence of changes:
1. Add new feature to allow detection of files based on file content (format detector).
2. Add tests for new feature.
3. Refactor `createReader` to utilize the new feature to improve functionality.

Nowhere in those steps do I think we are restructuring for the sole reason of testing existing features.

> If the code-under-test is completely isolated from the business logic (for example: ObjectPool), then (1) is not true, and it makes sense to move this code to separate header/source files.

How do you define complete isolation? To not depend on anything? To not be related to anything? To be reusable?
IMO, the format detector is isolated enough to be justified:
- It does not depend on any business logic. Business logic depends on it. Whether we create `PcapDevice` or `PcapNGDevice` is completely independent from the precise way of how we determine the format, only that we do.
- Unless the piece of code is unused, I don't think you can get a completely isolated piece of code, it it is supposed to not be related to anything.
- The format detector is reusable. It currently might not be needed to be reused anywhere else, but that does not stop it from being reused. I can copy + paste or include the implementation into a program that loops through all files and counts how many Pcap files are in a folder with zero changes to the detector's implementation.

> What I meant is that CaptureFileFormatDetector is not part of the public API, but rather an internal class

It is an internal class. The way I think about it, the collection of all internal classes are a micro library that the public API's implementation depends on and should be validated independently if possible with minimal complications. Having a separate header + source for a class is not a large complication.

> Again - the main thing that guides me is: do we need this refactoring just for the sake of testing? The answer in this case is "yes".

As I said in my first paragraph, the way I see it, this is not refactoring of existing code just for the sake of testing, but deciding how a new independent feature will be merged, alongside all its relevant tests to validate its integrity.

> Regarding the example you gave: if bar is completely isolated from Foo and may be reusable in other places (like ObjectPool), then sure - we can move bar to a separate header/source file. But if bar is an internal implementation of Foo, and we have a good enough way to test doFooImpl / bar using the public method doFoo then I'd go with this approach

Again, what do you define as "completely isolated"? For "may be reusable in other places", I think I answered that previously with the fact that I can 100% reuse the format detector wherever I need to differentiate between Pcap, PcapNG, Snoop files with minimal to no changes.

The only thing I might change is that `PcapNGZst` should return `ZstArchive` and the `Zst` to `PcapNG` folding should be done in the business logic of `createReader`, as it is not related to the detection of the format of the active file.

I think having that change will eliminate any mention of the necessary business logic in relation to `createReader` from the format detector internals, making it a fully independent module, similar to `ObjectPool`.

seladb replied:
> This is why I labeled the PR as enhancement, not refactor.
> Nowhere in those steps do I think we are restructuring existing code for the sole reason of testing existing features.

If it wasn't for the sake of testing - would you include `CaptureFileFormatDetector.h` as a separate file or have its implementation within `PcapFileDevice.cpp`? I think I know the answer 🙂 because it was initially inside `PcapFileDevice.cpp` and we only extracted it to a separate file for the sake of the tests...

> How do you define complete isolation? To not depend on anything? To not be related to anything? To be reusable?

I think that in most cases - a piece of infrastructure code that is not related to any specific business logic and is reusable. Most of what we have in Common++ is like that. But `CaptureFileFormatDetector` is very specific and will only be used in the pcap reader file

> It is an internal class. The way I think about it, the collection of all internal classes are a micro library that the public API's implementation depends on and should be validated independently if possible with minimal complications. Having a separate header + source for a class is not a large complication.

I agree it's not a large complication, but we almost never do it in PcapPlusPlus, and if we do, we need to have a good reason for it. Testing could be a good reason, but in this case the same test could be run on `createReader` even though the abstraction is not ideal

Dimi1010 replied:
> The logic in createReader does an educated guess, not a bullet-proof validation

Curious why wouldn't you want that?

I would reason that having a better validation sequence inside the factory would make for cleaner UX due to less boilerplate needed by the user?

```cpp
auto dev = IFileDevice::tryCreateReader("filePath");
if (!dev)
{
  // User has to handle error here.
}

// User has to open device.
bool res = dev.open();
if (!res)
{
  // User also has to handle error here.
  // Zst particularly may fail here if contents are not pcapng.
}

// Use device here.
```

Having the integrated open / validation would allow single line, before use. I don't see much use cases where you would want to create a device reader and *not* open to read from it, no? It also fits with the RAII methodology of avoiding a 2-stage init where possible.

```cpp
auto dev = IFileDevice::tryCreateReader("filePath");
if(!dev)
{
  // Failed to create device.
  // Note: No need for second boilerplate error handler prior to use.
}

// Use device here.
```

The live devices need `open()` because they are created by the runtime at startup.
File devices don't need to have that limitation since they are entirely created by the user.

> I don't think we want to unpack the Zstd archive just to see if it's valid. We have the pacpng library that does that.

Yes, but I am unsure if it gives a precise error message of what went wrong or just a generic failure error.

> Otherwise we can argue that checking the magic numbers is not enough - why not validate the entire pcap / pcapng file?

Which is as simple as calling `open()` inside the factory function, no? As you said, the backend already does validation, so why not reuse it for the factory validation?

Dimi1010 replied:
> Extracting the archived file just to verify the format is wasteful and might take a long time, especially if it's called on large files.

There is no need to extract the entire file. ZST compression works on independent frames allowing frame-by-frame (streaming) decompression. We only need to decompress the first frame to read the magic number to validate that the archive contents appear to be PcapNG. How large the total file is is irrelevant, as the frames are fixed size.

Incidentally, frame-by-frame is also how LightPcapNG reads the ZST archive. It decompresses a frame, reads the fully decompressed PcapNG records in it, and decompresses the next frame if needed.

> If this is indeed the case, maybe we need to fix the LightPcapNg code?

Which is C code, and makes it much harder to output a readable error. Not to mention we need to deal with passing that error up the stack.

> Not necessarily - as far as I know open() checks mostly the header and doesn't go over the rest of the file, so a user can open a file with a correct header but corrupted data and reading the file will fail

But it will still have passed `open()`. My idea isn't that `createReader` should validate that everything is correct. It is that it should validate just enough to guarantee that the returned device can successfully pass an `open()` call. The device might even be returned already opened and ready for reading, reducing the user side boilerplate.

There is no reason to return a device that can't even be opened, since the user can't do anything with it. It just adds more boilerplate as the user has to do the error handling twice.

If the records afterwards are corrupted at some point, the read should fail when the corrupted data is reached.

Dimi1010 replied:
> I asked ChatGPT, and there's indeed a way to read just the header using ZSTD_getFrameHeader - is this what you're referring to?

Not really, but similar. From what I understand, the ZST archive has the following format:
`[ZST Header] [CBlock (Compressed Block) 0] [CBlock 1] [CBlock 2] ... [CBlock N]`

Those blocks need to be decompressed sequentially starting from `CBlock 0`. (I was wrong earlier when I mentioned they were completely independent). The thing is, the magic number of the inner file is at the start of `CBlock 0`, so we don't need to decompress any of the other blocks. ZSTD provides a [streaming API](https://facebook.github.io/zstd/zstd_manual.html#Chapter7). We can be use that to only decompress `CBlock 0` to fetch the magic number of the compressed content and stop.

> Still, I'd be in favor of fixing it in LightPcapNg. In any case - PcapPlusPlus doesn't call Zstd API directly, everything goes through LightPcapNg so I think the right way would be to fix or add it in LightPcapNg

It is possible to add the check to light_pcapng. It would need to do a check around the time it tries to read the header block. My question is how to communicate that error condition to the pcpp layer? From what I can see in `light_pcapng_open_read`, if it fails to open the error message is outputted as `fprint(...)` assertion messages and the return value is `nullptr`. I suppose we can do something similar, but then the pcpp layer gets a generic pass / fail, and not really a reason why it failed.

> This is one way to look at it, another approach could say - createReader should do its best to guess the right reader, and whether it can be opened or not is a question for open(). I'm not sure which approach is better so I would vote for the simplest option 🙂

Both approaches are valid. I think the main decision point are the two questions:
- In what situation do you want to make a device and not open it? I think that would be a pretty rare occasion.
- What can you do with a device that can't open? I think the answer here is nothing, except to destroy it, which raises the question of why return such a device in the first place?

I updated the `createReader` procedure in ae9caa882ff9ea1f19996670b513796843017350 to include a `openDevice` parameter. I think that would do for a factory procedure for now? Users would have the option of keeping the old behaviour and manually opening the device later, but for the general case it is handled for them and they can immediately begin reading from it for ease of use.

 What do you think?

seladb replied:
> It is possible to add the check to light_pcapng. It would need to do a check around the time it tries to read the header block. My question is how to communicate that error condition to the pcpp layer? From what I can see in `light_pcapng_open_read`, if it fails to open the error message is outputted as `fprint(...)` assertion messages and the return value is `nullptr`. I suppose we can do something similar, but then the pcpp layer gets a generic pass / fail, and not really a reason why it failed.

This is a valid question, but maybe PcapPlusPlus doesn't need to know the exact reason - it only needs to know that it failed - `fprint` will print the reason like `GetLastError` used in many libraries



> Both approaches are valid. I think the main decision point are the two questions:
> 
>     * In what situation do you want to make a device and not open it? I think that would be a pretty rare occasion.
> 
>     * What can you do with a device that can't open? I think the answer here is nothing, except to destroy it, which raises the question of why return such a device in the first place?
> 
> 
> I updated the `createReader` procedure in [ae9caa8](https://github.com/seladb/PcapPlusPlus/commit/ae9caa882ff9ea1f19996670b513796843017350) to include a `openDevice` parameter. I think that would do for a factory procedure for now? Users would have the option of keeping the old behaviour and manually opening the device later, but for the general case it is handled for them and they can immediately begin reading from it for ease of use.
> 
> What do you think?

This is generally true with out API - usually when creating a device the next thing to do is opening it. I originall did it to avoid throwing exceptions or handling validations in the c'tor. I'd consider `createReader` as a "smart constructor" that decides on the object to return, but the rest of the API should stay consistent -> the next action would be calling `open()`

Dimi1010 replied:
Sure. Reverted the `create` methods to match the old behaviour of not returning opened devices.

Dimi1010 commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -29,43 +62,210 @@ namespace pcpp
+				return CaptureFileFormat::Unknown;
+			}
+
+			return formatMapping[std::distance(pcapMagicNumbers.begin(), it) / 2];
```
I admit its not really the most readable, but it was a compromise.

I also thought of having 1 array of `pair<uint32_t, CaptureFormat>`, but the current implementation has `std::find` can get compiled to a SIMD search, due to all the `uint32_t` magic values being tightly packed. `std::find_if` does not have that optimization.

I can wrap the format mapping in a lambda `getFormatMapping()` that gets defined near the array definitions if you want, but I think the comments are clear enough? I am also unsure if the optimized will inline the lambda but it is likely it will?

Dimi1010 replied:
Thought about it some more. Sure, I will update it to the if statement version.

Dimi1010 replied:
Updated 181a8b4be1ee9d14f325d2998b1dbc33426529d6

Dimi1010 committed:
Updated detectPcap selection.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Extracted capture format detector to remove it from publicly available headers.

Dimi1010 committed:
Fix includes.

Dimi1010 committed:
Removed duplicate files from tracking.

Dimi1010 committed:
Lint

Dimi1010 commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -26,6 +26,204 @@ class FileReaderTeardown
+	// Correct format
+	constexpr const char* PCAPNG_FILE_PATH = "PcapExamples/file_heuristics/pcapng-example.pcapng";
+	// Correct format, wrong extension
+	constexpr const char* PCAPNG_AS_PCAP_FILE_PATH = "PcapExamples/file_heuristics/pcapng-with-pcap-ext.pcapng.pcap";
```
Fixed e275950b054b1ecb620cdc48a697c3f3e80ec818

Dimi1010 commented on Tests/Pcap++Test/PcapExamples/file_heuristics/pcapng-with-pcap-ext.pcapng.pcap:
Fixed e275950b054b1ecb620cdc48a697c3f3e80ec818

Dimi1010 committed:
Trimmed pcapng sample.

Dimi1010 commented on Tests/Pcap++Test/PcapExamples/file_heuristics/pcapng-example.pcapng:
Fixed 54f7baee9a28cd9765ebcd69507ffc890ce79d26

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Change PcapNGZst to ZstArchive. Zst to PcapNG branch folding is done in `createReader` instead of having the Format detector assume that is what is intended.

Dimi1010 committed:
Added separate format value for "modified" pcap to separate the format detector from libpcap behaviour.

Dimi1010 committed:
Docs fix.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Add automatic open functionality to `createReader` factory to mimic RAII initialization.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Update docstring.

Dimi1010 committed:
Fix exception message assert.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Refactored format tests to utilize the createReader factory.

Dimi1010 committed:
Fix nanoprecision test issues.

Dimi1010 committed:
Remove openDevice flag. Update create procedure to avoid exceptions on tryCreateDevice.

Dimi1010 committed:
Docs update + Lint

Dimi1010 committed:
Docs fix.

Dimi1010 committed:
Lint.

seladb commented:
@Dimi1010 I'm working on parsing pcap files without libpcap: https://github.com/seladb/PcapPlusPlus/pull/2034
Maybe we can rework this PR after my PR is merged?

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

# Conflicts:
#	Pcap++/src/PcapFileDevice.cpp
#	Tests/Pcap++Test/Tests/FileTests.cpp

Dimi1010 committed:
Remove nano support checks as it should always be supported by the internal parser.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -24,162 +42,371 @@ namespace pcpp
+		};
+
+		/// @brief Heuristic file format detector that scans the magic number of the file format header.
+		class CaptureFileFormatDetector
```
Since we're not parsing all formats (maybe except Zstd) in PcapPlusPlus, we can reuse the logic we already have. Maybe it can run the `open()` method (or extract a portion of it) for each reader type until it can to find the right type?

Dimi1010 replied:
Hmm, it should be possible. It will require expanding the `CaptureFileFormatDetector` a bit. Currently it only returns the format, but pcap for instance uses the magic number to also detect native or swapped byte order.

Depending on how specific we want to get it might involve a double read of the magic number, once by the format detector and once during the actual file header structure read. Impact should be minimal tho, as `fstream` is buffered by default.

Dimi1010 replied:
I suppose that is also an option. Although that does scatter all the format checks across all devices?
There are also some devices that handle other formats that don't really map as intuitively to the device name, e.g. `ztd` in PcapNG. I wanted the format detector to be a centralized format checker to avoid scattering the code all over.

seladb replied:
I think that each file device should own its own format check, especially since we no longer use libpcap. We do it anyway in the different file device. PcapNG with/without Zstd is a bit of an exception because we use `LightPacpNg` to parse these files - but I still think it'd be better if `PcapNgFileReaderDevice` owns the format detection of PcapNG and Zstd

Dimi1010 replied:
> I think that each file device should own its own format check, especially since we no longer use libpcap. We do it anyway in the different file device. PcapNG with/without Zstd is a bit of an exception because we use `LightPacpNg` to parse these files - but I still think it'd be better if `PcapNgFileReaderDevice` owns the format detection of PcapNG and Zstd

@seladb minor note to that. I did look over how it might happen. It would lead to some code duplication in `PcapReader` and `PcapWriter` (append mode), since they both need to validate a pcap file magic signature + byte order.

Also do we really need to tie them to the public API? The `isPcapFile` in the example is part of the public API while the current implementation is entirely private. The main API entry point is `createReader` anyway, so the user shouldn't really need all the `isXXXFile`, necessarily?

seladb replied:
> @seladb minor note to that. I did look over how it might happen. It would lead to some code duplication in `PcapReader` and `PcapWriter` (append mode), since they both need to validate a pcap file magic signature + byte order.

Where would be the code duplication in this case?



> Also do we really need to tie them to the public API? The `isPcapFile` in the example is part of the public API while the current implementation is entirely private. The main API entry point is `createReader` anyway, so the user shouldn't really need all the `isXXXFile`, necessarily?

I agree it's not great, but is there a workaround which will keep the parsing logic in each class and make it available for the reader creator? 🤔

Dimi1010 replied:
> > @seladb minor note to that. I did look over how it might happen. It would lead to some code duplication in `PcapReader` and `PcapWriter` (append mode), since they both need to validate a pcap file magic signature + byte order.
> 
> Where would be the code duplication in this case?

See #2082. That PR solves the duplication.


> > Also do we really need to tie them to the public API? The `isPcapFile` in the example is part of the public API while the current implementation is entirely private. The main API entry point is `createReader` anyway, so the user shouldn't really need all the `isXXXFile`, necessarily?
> 
> I agree it's not great, but is there a workaround which will keep the parsing logic in each class and make it available for the reader creator? 🤔 

Eh. Not really. :/ Tho, if it isn't going to be public api, it doesn't nessesarily need to be a member function exposed in the headers, no? It can alternatively be a quasi adjacent free function helper in the cpp.

seladb replied:
> See #2082. That PR solves the duplication.

I saw this PR, not sure we really need to solve it, please see my comment there



> Eh. Not really. :/ Tho, if it isn't going to be public api, it doesn't nessesarily need to be a member function exposed in the headers, no? It can alternatively be a quasi adjacent free function helper in the cpp.

I don't think it's so bad if it's part of the public API, even though ideally it shouldn't be. We have such cases in other places as well

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Extend format detector to capture byte order from magic number.

Dimi1010 committed:
Add initial merge logic.

Dimi1010 committed:
Add methods to check magic numbers directly.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Comment out unused variables.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

# Conflicts:
#	Pcap++/src/PcapFileDevice.cpp

Dimi1010 committed:
Add helper to convert format to pcap precision.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

# Conflicts:
#	Pcap++/src/PcapFileDevice.cpp

Dimi1010 committed:
Simplify code.

Dimi1010 committed:
Fixed errors.

Dimi1010 committed:
Disable pcap "modified" file format detection tests.

Dimi1010 committed:
Lint

Dimi1010 commented:
Ok, somewhat reworked the internal details on this.

`CaptureFileFormatDetector` was dropped along with all the `is[Pcap|PcapNG|Snoop|Zst]File` methods.

They were replaced with more basic internal free function helpers `is***Magic` which take `uint32_t` (`uint64_t` for snoop) for the suspected magic value and return format and byte order (if it can be determined).

That allows them to be reused in both `***Device::open` checks and `detectFileFormat` without overlap.

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

Dimi1010 committed:
Merge branch 'dev' into feature/heuristic-file-selection

# Conflicts:
#	Pcap++/header/PcapFileDevice.h
#	Pcap++/src/PcapFileDevice.cpp

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -269,6 +538,143 @@ namespace pcpp
+		}
+	}
+
+	std::unique_ptr<IFileReaderDevice> IFileReaderDevice::tryCreateReader(const std::string& fileName)
```
Shouldn't `tryCreateReader` simply call `createReader()`, catch the exception and return `nullptr`?

I think it's safe to assume that reading a file is not a part of the fast-path and the overhead of `try..catch` is less significant comparing to reading and analyzing the file

Dimi1010 replied:
It did do that initially. I reworked it to avoid the exception, but I can revert it if you think it will be simpler to maintain in the long run.

seladb replied:
Yes, calling `createReader()` will require less code and be easier to maintain so I'd recommend making this change

Dimi1010 replied:
Fixed 61187f976ae70a0de983f13162675eced4b93406

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -36,180 +42,443 @@ namespace pcpp
+		/// The function optionally detects the byte order of the file if it can be determined by the magic number.
+		/// The byte order is not updated if no supported format is detected.
+		///
+		/// @param[in] content A stream that contains the file content.
```
This is an internal method so doxygen is not required. However if we want to keep it, we can probably change `content` to `magic`

Dimi1010 replied:
Yeah, thanks for the catch. I suppose it isn't required. I just prefer to write at least some simple docs even for internal methods unless it is something super simple.

Dimi1010 replied:
Fixed 296d78f793f3d2500740692ef828e96f79bff292

seladb commented on Pcap++/src/PcapFileDevice.cpp:
```diff
@@ -36,180 +42,443 @@ namespace pcpp
+				// Libpcap 0.9.1 and later support reading a modified pcap format that contains an extended header.
+				// Format reference: https://wiki.wireshark.org/Development/LibpcapFileFormat#modified-pcap
+				0xa1'b2'cd'34,  // Alexey Kuznetzov's modified libpcap format
+				0x34'cd'b2'a1,  // Alexey Kuznetzov's modified libpcap format (byte-swapped)
```
Maybe we can add these as constexpr also?

Dimi1010 replied:
Sure. Those are also interesting since they are technically `Pcap` files, but I think some of the headers are extended, so we can't actually parse them with the pcpp parser (libpcap could). I am not sure how widely they are used tho?

Dimi1010 replied:
Fixed 62a0ecc36deb24b2c66a5cae9264f5773370a7b9

seladb commented on Tests/Pcap++Test/PcapExamples/file_heuristics/temp/.gitignore:
Since we only write `.tmp`, `.dat` and one pcap file, we can add this `.gitignore` in `Tests/Pcap++Test/PcapExamples/file_heuristics` and put `*.tmp`, `*.dat` and the specific pcap file. I think it's nicer than having an empty `temp/` folder

Dimi1010 replied:
Fixed 0993ff79fbbd3e2357143c5acb2c7a48f38bc685

Ended up dropping the `.dat` to keep things simple. Renamed the `.dat` file to `.tmp` too.

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Fixed doxygen.

Dimi1010 committed:
Change Kuznetzov pcap magic number to constexpr.

Dimi1010 committed:
Change all temporary files to have `.tmp` extension. Remove `/temp/` subdirectory.

Dimi1010 committed:
Lint

seladb commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -30,6 +30,282 @@ class FileReaderTeardown
+		PTF_ASSERT_NOT_NULL(dynamic_cast<pcpp::PcapFileReaderDevice*>(dev.get()));
+	}
+
+	if (pcpp::PcapFileReaderDevice::isNanoSecondPrecisionSupported())
```
I don't think we need to check `isNanoSecondPrecisionSupported()` anymore since it's a deprecated method

Dimi1010 replied:
Fixed 8b7f7eb8c3185f3bb5ae2b19de3160d0e626f660

seladb commented on Tests/Pcap++Test/Tests/FileTests.cpp:
```diff
@@ -30,6 +30,282 @@ class FileReaderTeardown
+		PTF_ASSERT_NOT_NULL(dynamic_cast<pcpp::PcapNgFileReaderDevice*>(dev.get()));
+	}
+
+	if (pcpp::PcapNgFileReaderDevice::isZstdSupported())
```
I think we can remove this `if` statement also. The zstd file type checker works even if PcapPlusPlus is compiled without zstd support

Dimi1010 replied:
The file type checker works, but the device factory throws a `std::runtime_error` if a zstd file is detected and the library does not have zstd support, because the device would be invalid. Therefor the check is needed, otherwise the test fails with exception.

If we did the tests of the file type checker independently of device creation that wouldn't have mattered, but since we are shimming them through the device factory...

seladb replied:
Oh you're right, my bad 🤦

seladb reviewed (approved):
Please see a few minor comments, otherwise LGTM

Dimi1010 committed:
Merge remote-tracking branch 'upstream/dev' into feature/heuristic-file-selection

Dimi1010 committed:
Revert `tryCreateReader` to use `createReader`

Dimi1010 committed:
Remove conditional branch that always evaluates to true.

Dimi1010 committed:
Lint
